// vertex-ai-mcp-server/build/tools/change_text_tone.js
import { McpError, ErrorCode } from "@modelcontextprotocol/sdk/types.js";
import { modelIdPlaceholder } from "./tool_definition.js";
export const changeTextToneTool = {
    name: "change_text_tone",
    description: `Rewrites a given block of text to match a specified target tone (e.g., formal, casual, humorous). Uses the configured Vertex AI model (${modelIdPlaceholder}). Requires 'text' and 'targetTone'.`,
    inputSchema: { type: "object", properties: { text: { type: "string", description: "The text content to rewrite." }, targetTone: { type: "string", description: "The desired tone for the rewritten text (e.g., 'formal', 'casual', 'optimistic')." } }, required: ["text", "targetTone"] },
    buildPrompt: (args, modelId) => {
        const { text, targetTone } = args;
        if (typeof text !== "string" || !text || typeof targetTone !== "string" || !targetTone)
            throw new McpError(ErrorCode.InvalidParams, "Missing 'text' or 'targetTone'.");
        return {
            systemInstructionText: `You are a writing assistant. Rewrite the text in a ${targetTone} tone. Format as Markdown.`,
            userQueryText: `Text:\n${text}`,
            useWebSearch: false,
            enableFunctionCalling: false
        };
    }
};
