// vertex-ai-mcp-server/build/tools/explain_regex.js
import { McpError, ErrorCode } from "@modelcontextprotocol/sdk/types.js";
import { modelIdPlaceholder } from "./tool_definition.js";
export const explainRegexTool = {
    name: "explain_regex",
    description: `Provides a step-by-step explanation of a given regular expression pattern. Uses the configured Vertex AI model (${modelIdPlaceholder}). Requires 'regex'.`,
    inputSchema: { type: "object", properties: { regex: { type: "string", description: "The regular expression pattern string to explain." } }, required: ["regex"] },
    buildPrompt: (args, modelId) => {
        const { regex } = args;
        if (typeof regex !== "string" || !regex)
            throw new McpError(ErrorCode.InvalidParams, "Missing 'regex'.");
        return {
            systemInstructionText: `You are a regex expert. Break down and explain the regex pattern step-by-step. Format as Markdown.`,
            userQueryText: `Regex: ${regex}`,
            useWebSearch: false,
            enableFunctionCalling: false
        };
    }
};
