// vertex-ai-mcp-server/build/tools/extract_keywords.js
import { McpError, ErrorCode } from "@modelcontextprotocol/sdk/types.js";
import { modelIdPlaceholder } from "./tool_definition.js";
export const extractKeywordsTool = {
    name: "extract_keywords",
    description: `Extracts the main keywords or topics from a block of text. Uses the configured Vertex AI model (${modelIdPlaceholder}). Requires 'text'.`,
    inputSchema: { type: "object", properties: { text: { type: "string", description: "The text content to extract keywords from." } }, required: ["text"] },
    buildPrompt: (args, modelId) => {
        const { text } = args;
        if (typeof text !== "string" || !text)
            throw new McpError(ErrorCode.InvalidParams, "Missing 'text'.");
        return {
            systemInstructionText: `You are a data analysis assistant. Extract the main keywords or topics from the text. Output as a JSON array of strings.`,
            userQueryText: `Text:\n${text}`,
            useWebSearch: false,
            enableFunctionCalling: false
        };
    }
};
