// vertex-ai-mcp-server/build/tools/generate_regex.js
import { McpError, ErrorCode } from "@modelcontextprotocol/sdk/types.js";
import { modelIdPlaceholder } from "./tool_definition.js";
export const generateRegexTool = {
    name: "generate_regex",
    description: `Generates a regular expression pattern based on a natural language description. Uses the configured Vertex AI model (${modelIdPlaceholder}). Requires 'description'.`,
    inputSchema: { type: "object", properties: { description: { type: "string", description: "Natural language description of the pattern the regex should match." } }, required: ["description"] },
    buildPrompt: (args, modelId) => {
        const { description } = args;
        if (typeof description !== "string" || !description)
            throw new McpError(ErrorCode.InvalidParams, "Missing 'description'.");
        return {
            systemInstructionText: `You are a regex generation expert. Create a functional regex based *only* on the description. Output *only* the raw regex pattern.`,
            userQueryText: `Description: ${description}`,
            useWebSearch: false,
            enableFunctionCalling: false
        };
    }
};
