// vertex-ai-mcp-server/build/tools/generate_code_summary.js
import { McpError, ErrorCode } from "@modelcontextprotocol/sdk/types.js";
import { modelIdPlaceholder } from "./tool_definition.js";
export const generateCodeSummaryTool = {
    name: "generate_code_summary",
    description: `Creates a concise, high-level summary of a given code snippet's purpose and functionality. Uses the configured Vertex AI model (${modelIdPlaceholder}). Requires 'code' and 'language'.`,
    inputSchema: { type: "object", properties: { code: { type: "string", description: "The code snippet to summarize." }, language: { type: "string", description: "The programming language of the code snippet." } }, required: ["code", "language"] },
    buildPrompt: (args, modelId) => {
        const { code, language } = args;
        if (typeof code !== "string" || !code || typeof language !== "string" || !language)
            throw new McpError(ErrorCode.InvalidParams, "Missing 'code' or 'language'.");
        return {
            systemInstructionText: `You are a code analysis assistant. Provide a concise, high-level summary of the ${language} code. Format as Markdown.`,
            userQueryText: `Code:\n\`\`\`${language}\n${code}\n\`\`\``,
            useWebSearch: false,
            enableFunctionCalling: false
        };
    }
};
