// vertex-ai-mcp-server/build/tools/generate_sql_query.js
import { McpError, ErrorCode } from "@modelcontextprotocol/sdk/types.js";
import { modelIdPlaceholder } from "./tool_definition.js";
export const generateSqlQueryTool = {
    name: "generate_sql_query",
    description: `Generates a SQL query based on a natural language description. Can optionally use provided database schema context. Uses the configured Vertex AI model (${modelIdPlaceholder}). Requires 'description'.`,
    inputSchema: { type: "object", properties: { description: { type: "string", description: "Natural language description of the desired SQL query." }, schemaContext: { type: "string", description: "Optional string containing database schema information (e.g., CREATE TABLE statements)." } }, required: ["description"] },
    buildPrompt: (args, modelId) => {
        const { description, schemaContext } = args;
        if (typeof description !== "string" || !description)
            throw new McpError(ErrorCode.InvalidParams, "Missing 'description'.");
        const contextStr = schemaContext ? ` DB Schema Context:\n${schemaContext}` : '';
        return {
            systemInstructionText: `You are a SQL expert. Generate a correct SQL query based *only* on the description and optional schema context. Output *only* the raw SQL query.`,
            userQueryText: `Description: ${description}.${contextStr}`,
            useWebSearch: false,
            enableFunctionCalling: false
        };
    }
};
