// vertex-ai-mcp-server/build/tools/generate_commit_message.js
import { McpError, ErrorCode } from "@modelcontextprotocol/sdk/types.js";
import { modelIdPlaceholder } from "./tool_definition.js";
export const generateCommitMessageTool = {
    name: "generate_commit_message",
    description: `Generates a conventional commit message based on a provided code diff. Uses the configured Vertex AI model (${modelIdPlaceholder}). Requires 'diff'.`,
    inputSchema: { type: "object", properties: { diff: { type: "string", description: "A string containing the code changes (output of 'git diff')." } }, required: ["diff"] },
    buildPrompt: (args, modelId) => {
        const { diff } = args;
        if (typeof diff !== "string" || !diff)
            throw new McpError(ErrorCode.InvalidParams, "Missing 'diff'.");
        return {
            systemInstructionText: `You are a commit message assistant. Generate a concise and informative conventional commit message based *only* on the provided diff. Output *only* the commit message text.`,
            userQueryText: `Diff:\n\`\`\`diff\n${diff}\n\`\`\``,
            useWebSearch: false,
            enableFunctionCalling: false
        };
    }
};
