// vertex-ai-mcp-server/build/tools/summarize_text.js
import { McpError, ErrorCode } from "@modelcontextprotocol/sdk/types.js";
import { modelIdPlaceholder } from "./tool_definition.js";
export const summarizeTextTool = {
    name: "summarize_text",
    description: `Summarizes a given block of text to a specified length (short, medium, detailed). Uses the configured Vertex AI model (${modelIdPlaceholder}). Requires 'text'.`,
    inputSchema: { type: "object", properties: { text: { type: "string", description: "The text content to summarize." }, length: { type: "string", enum: ["short", "medium", "detailed"], description: "Optional desired summary length ('short', 'medium', 'detailed'). Defaults to 'medium'." } }, required: ["text"] },
    buildPrompt: (args, modelId) => {
        const { text, length } = args;
        if (typeof text !== "string" || !text)
            throw new McpError(ErrorCode.InvalidParams, "Missing 'text'.");
        const lengthStr = length ? ` Target length: ${length}.` : '';
        return {
            systemInstructionText: `You are a text summarization assistant. Summarize the provided text accurately and concisely.${lengthStr} Format as Markdown.`,
            userQueryText: `Text:\n${text}`,
            useWebSearch: false,
            enableFunctionCalling: false
        };
    }
};
